kernel/kernel_drivers.go

```diff
@@ -0,0 +1,146 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package kernel
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// For testing purposes
+var osSymlink = os.Symlink
+
+// KernelVersionFromPlaceInfo returns the kernel version for a mounted kernel
+// snap (this would be the output if "uname -r" for a running kernel).
+func KernelVersionFromPlaceInfo(mountPoint string) (string, error) {
+	systemMapPathPrefix := filepath.Join(mountPoint, "System.map-")
+	matchPath := systemMapPathPrefix + "*"
+	matches, err := filepath.Glob(matchPath)
+	if err != nil {
+		// could be only ErrBadPattern, should not really happen
+		return "", fmt.Errorf("internal error: %w", err)
+	}
+	if len(matches) != 1 {
+		return "", fmt.Errorf("unexpected number of matches (%d) for glob %s", len(matches), matchPath)
+	}
+	version := strings.TrimPrefix(matches[0], systemMapPathPrefix)
+	if version == "" {
+		return "", fmt.Errorf("kernel version not set in %s", matches[0])
+	}
+	return version, nil
+}
+
+func createFirmwareSymlinks(origin, dest string) error {
+	entries, err := os.ReadDir(origin)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range entries {
+		lpath := filepath.Join(dest, node.Name())
+		origPath := filepath.Join(origin, node.Name())
+		switch node.Type() {
+		case 0, fs.ModeDir:
+			// Create link for regular files or directories
+			os.Symlink(origPath, lpath)
+		case fs.ModeSymlink:
+			// Replicate link (it should be relative)
+			lpath := filepath.Join(dest, node.Name())
+			dest, err := os.Readlink(origPath)
+			if err != nil {
+				return err
+			}
+			os.Symlink(dest, lpath)
+		default:
+			return fmt.Errorf("unexpected file type: %s", node.Type())
+		}
+	}
+
+	return nil
+}
+
+// BuildKernelDriversTree creates a tree in
+// /var/lib/snapd/kernel/<ksnapName>/<rev> with subfolders can be bind-mounted
+// on boot to /usr/lib/{modules,firmware}. This tree contains files from the
+// kernel snap mounted on kernelMount, as well as symlinks to it.
+// TODO this will be extended to consider kernel-modules components.
+func BuildKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string) (err error) {
+	// Create directory tree
+	revStr := rev.String()
+	treeRoot := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "kernel", ksnapName, revStr)
+	kversion, err := KernelVersionFromPlaceInfo(kernelMount)
+	if err != nil {
+		return err
+	}
+	modsRoot := filepath.Join(treeRoot, "lib", "modules", kversion)
+	firmwareRoot := filepath.Join(treeRoot, "lib", "firmware")
```

Comment:
Since we'd be bind mounting respective directories, you don't need the complete prefix inside the tree

```suggestion
	modsRoot := filepath.Join(treeRoot, "modules", kversion)
	firmwareRoot := filepath.Join(treeRoot, "firmware")
```